### PR TITLE
[DO NOT MERGE] For Swift version 3 / 4, treat IUOs in illegal positions as Optionals.

### PR DESCRIPTION
**Pull request by rudkx**
Emit a warning if '!' appears in a position where it was banned in
SE-0054 and we're in Swift version < 5. We'll treat these as Optionals
instead and attempt to continue compilation.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

File: test/Interpreter/FunctionConversion.swift [hunk 1]
```diff
@@ -84,19 +84,19 @@ func a1(s: AddrOnly?) -> AddrOnly {
...
   let g11: (Trivial) -> Trivial? = t1
-  let g12: (Trivial!) -> Trivial? = t1
+  let g12: (Trivial?) -> Trivial? = t1
 
   expectEqual(22, g11(Trivial(n: 11))?.n)
   expectEqual(24, g12(Trivial(n: 12))?.n)
 
   let g21: (Loadable?) -> Loadable? = l1
-  let g22: (Loadable!) -> Loadable? = l1
+  let g22: (Loadable?) -> Loadable? = l1
```
Comment (slavapestov) [inline test/Interpreter/FunctionConversion.swift]: This line is redundant now

File: test/Interpreter/FunctionConversion.swift [hunk 2]
```diff
@@ -84,19 +84,19 @@ func a1(s: AddrOnly?) -> AddrOnly {
...
 
   expectEqual(42, g21(Loadable(n: 21))?.n)
   expectEqual(44, g22(Loadable(n: 22))?.n)
 
   let g31: (AddrOnly?) -> AddrOnly? = a1
-  let g32: (AddrOnly!) -> AddrOnly? = a1
+  let g32: (AddrOnly?) -> AddrOnly? = a1
```
Comment (slavapestov) [inline test/Interpreter/FunctionConversion.swift]: This line is redundant now

File: test/Migrator/wrap_optional.swift [hunk 3]
```diff
@@ -36,19 +36,15 @@ class MyExtraCities : ExtraCities {
 }
 
 typealias IntAnd<T> = (Int, T)
-class Outer {
-  typealias Inner = (String?, String) -> String!
-}
 
 class MyExtraCitiesWithAliases : ExtraCities {
-  func blibli(x: Outer.Inner) {}
+  func blibli(x: (String?, String) -> String!) {}
```
Comment (slavapestov) [inline test/Migrator/wrap_optional.swift]: Why is this allowed but the typealias form is not? I would expect a function type would never be able to have an IUO return, only a function *declaration* would

File: validation-test/compiler_crashers_fixed/28638-anonymous-namespace-verifier-checksametype-swift-type-swift-type-char-const.swift [hunk 4]
```diff
@@ -5,5 +5,5 @@
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: %target-swift-frontend %s -emit-ir
-nil as Int!as?Int?!
+// RUN: %target-swift-frontend %s -emit-ir -verify
```
Comment (slavapestov) [inline validation-test/compiler_crashers_fixed/28638-anonymous-namespace-verifier-checksametype-swift-type-swift-type-char-const.swift]: Change this to 'not %target-swift-frontend %s -typecheck', and drop the -verify. We don't want validation tests to verify diagnostics

**Review by slavapestov — COMMENTED** [review]
Excellent!

File: test/Interpreter/FunctionConversion.swift (hunk 1 above)
Comment (rudkx) [reply test/Interpreter/FunctionConversion.swift]: Hmm, I thought I caught these and removed them. Thanks for pointing them out.

File: test/Migrator/wrap_optional.swift (hunk 3 above)
Comment (rudkx) [reply test/Migrator/wrap_optional.swift]: Yeah this was an oversight. This was getting caught at one point but I tweaked some code that resulted in this being accepted again and I didn't have a separate test confirming the diagnostic. I'll add one.

File: validation-test/compiler_crashers_fixed/28638-anonymous-namespace-verifier-checksametype-swift-type-swift-type-char-const.swift (hunk 4 above)
Comment (rudkx) [reply validation-test/compiler_crashers_fixed/28638-anonymous-namespace-verifier-checksametype-swift-type-swift-type-char-const.swift]: Yes, good point.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
Closing in favor of a new PR I will open soon.

**Pull request closed without merging**